auth/oidc.go

```diff
@@ -439,3 +514,182 @@ func SetURLQueryParam(strURL, name, value string) (string, error) {
 	uri.RawQuery = rawQuery.Encode()
 	return uri.String(), nil
 }
+
+// Run runs the provider metadata sync task in background.
+func (op *OIDCProvider) StartDiscoverySync() error {
+	op.stopDiscoverySyncOnce.Do(func() {
+		op.stopDiscoverySync = make(chan struct{})
+	})
+	duration, err := op.runDiscoverySync()
+	if err != nil {
+		return err
+	}
+	go func() {
+		for {
+			select {
+			case <-time.After(duration):
+				duration, err = op.runDiscoverySync()
+				if err != nil {
+					base.Errorf("openID Connect provider discovery sync ends up in error: %v", err)
+					return
+				}
+			case <-op.stopDiscoverySync:
+				base.Debugf(base.KeyAll, "Terminating OpenID Connect provider discovery sync")
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// sync polls the provider metadata from discovery endpoint and refresh the ID token verifier.
+func (op *OIDCProvider) runDiscoverySync() (time.Duration, error) {
+	discoveryURL := strings.TrimSuffix(op.Issuer, "/") + OIDCDiscoveryConfigPath
+	base.Debugf(base.KeyAuth, "Syncing provider metadata from: %s", base.UD(discoveryURL))
+	req, err := http.NewRequest(http.MethodGet, discoveryURL, nil)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error building new request for URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("unsuccessful response: %v", err)
+	}
+
+	var expiresAt time.Time
+	var ttl time.Duration
+	var ok bool
+	ttl, ok, err = cacheable(resp.Header)
+	if err != nil {
+		return 0, err
+	} else if ok && ttl > 0 {
+		expiresAt = time.Now().UTC().Add(ttl)
+	}
+
+	metadataChecksum, err := sha1Checksum(resp)
+	if op.metadataChecksum == metadataChecksum {
```

Comment:
This looks like it's always going to set the verifier on first use.  So the first time InitOIDCClient is run, it's going to set the verifier via DiscoverConfig, then turn around and set it again right away in StartDiscoverySync?  That approach has a couple of problems:
  1. It's inefficient.
  2. When using a standard discovery endpoint, we wanted to leverage the library for verifier creation.  This breaks that handling.

Reply:
As discussed, added changes to enforce the discovery sync to run just before the initialization of OIDC client